### DISCUSSION TITLE
SERVICE mechanism SPARQL-compliant?

### DISCUSSION THREAD

**Discussion by sebastianthelen** [SPARQL-Anything/sparql.anything#486] [Q&A]
Dear sparql.anything community, 

I don't know if this topic has already been addressed. The discussion in https://github.com/SPARQL-Anything/sparql.anything/discussions/388 seems to go in a similar direction, but couldn't answer my question. Maybe you can help me. 

Am I correct in saying that by overloading the SERVICE operator and supporting constructs like `SERVICE <x-sparql-anything:>`, sparql.anything is not fully SPARQL-compliant anymore? If this is true, it will prevent users from invoking sparql.anything from other endpoints via query federation, which I am considering doing for my project. Something like:

```
PREFIX xyz: <http://sparql.xyz/facade-x/data/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX fx: <http://sparql.xyz/facade-x/ns/>

SELECT *
WHERE {
    # local sparql.anything endpoint
    SERVICE <http://localhost:3000/sparql> { 
        SERVICE <http://x-sparql-anything> {
            fx:properties fx:location "https://www.w3schools.com/xml/note.xml" .
            fx:properties fx:triplifier "io.github.sparqlanything.xml.XMLTriplifier" .
            ?s ?p ?o .
        }
    }
}
```
In this case, sparql.anything could only be used as a standalone tool without the possibility of integrating it with other endpoints.

Thanks in advance for your help, 
Sebastian

**Comment by enridaga** [comment]
Hi Sebastian, and welcome! If I understand the scenario from your example, you want a sparql endpoint to send a SERVICE call to a sparql anything server that will include a sparql-anything service clause. Of course you can, this is the intended behaviour. What won't work is sending a service clause  that includes another sparql-anything service clause to a non-sparql-anything endpoint. The reason is obvious, the remote endpoint will not know how to process the sparql-anything service clause.
Now, I am not sure what you mean with "compliance" but that's probably beside the point.

**Comment by sebastianthelen** [comment]
Hi Enrico, 

thanks a lot for your feedback.

Turns out the initial query I posted had two issues:

- The actutal SPARQL Anything endpoint referenced in the outer SERVICE clause is exposed at <http://localhost:3000/sparql.anything> and not at <http://localhost:3000/sparql> as I was incorrectly assuming.
-   The inner SERVICE clause should call  `<x-sparql-anything:>` and not `<http://x-sparql-anything>.`

I fixed these two points and now the query works as expected. 

Again, thx a lot.
Sebastian

**Comment by justin2004** [comment]
@enridaga i think we can mark this as answered
